src/secontext.c

```diff
@@ -161,44 +164,123 @@ selinux_getfilecon(struct tcb *tcp, const char *path, char **secontext,
 	if (!proc_pid)
 		return -1;
 
-	int rc = -1;
-	char fname[PATH_MAX];
+	/*
+	 * The path may contain .../proc/self/... or .../proc/self<eos>.
+	 * We need to replace 'self' by the target PID, or else we would
+	 * resolve the path with 'self' being strace itself.
+	 * In some corner cases, it's impossible to guess so keep it as is if
+	 * we don't find any occurrence of /proc/self.
+	 * Finally there may be breakage if the pathname contains /proc/self
+	 * but has nothing to do with /proc, but it's impossible to tell and
+	 * unlikely to happen.
+	 */
+
+	int rc;
+	char fname[PATH_MAX + 1];
+	char buf[PATH_MAX + 1];
+
+	if (strlen(path) >= sizeof(fname))
+		return -1;
+	strcpy(fname, path);
+	int proc_self_substituted = 0;
+
+	/*
+	 * We replace all the occurrences of /proc/self[/] because we cannot
+	 * know how many there are (e.g. /proc/self/root/proc/self/root/foo).
+	 */
+	for (;;) {
+		char *s = strstr(fname, "/proc/self/");
+		if (s == NULL) {
+			/* Corner case: check if the path ends with /proc/self */
+			s = fname + strlen(fname) - sizeof("/proc/self") + 1;
+			if (strcmp("/proc/self", s))
```

Comment:
When the string in `fname` isn't long enough, this would result to an out of bounds read.